python/ql/test/library-tests/frameworks/gradio/source_test.ql

```diff
@@ -0,0 +1,5 @@
+import python
+import semmle.python.dataflow.new.RemoteFlowSources
+
+from RemoteFlowSource rfs
+select rfs
```

Comment:
This test can be made inline in the following way:
```suggestion
import python
import semmle.python.dataflow.new.RemoteFlowSources
import TestUtilities.InlineExpectationsTest
private import semmle.python.dataflow.new.internal.PrintNode

module SourceTest implements TestSig {
  string getARelevantTag() { result = "source" }

  predicate hasActualResult(Location location, string element, string tag, string value) {
    exists(location.getFile().getRelativePath()) and
    exists(RemoteFlowSource rfs |
      location = rfs.getLocation() and
      element = rfs.toString() and
      value = prettyNode(rfs) and
      tag = "source"
    )
  }
}

import MakeTest<SourceTest>
```
You will then have to adjust the test files by adding annotations (eg. line 15 of `source_test.py` becomes `def greet(name): # $ source=name`, the test failures will tell you what to add where) and making `source_test.expected` empty.

Reply:
Thank you! I added the annotations and fixed the test files in https://github.com/github/codeql/pull/16135/commits/d6acea1d0c97d06679cd33f53509ba4c4ea4040b. I'm unsure how to test the taint step though - is it fine as it is now?

Reply:
Yes that looks good. Testing the taint step is a bit more difficult because it is `private`. We could add provenance to it and test it that way, but we can leave this for later.